### PR TITLE
Typed Settings

### PR DESCRIPTION
**Pull request by NobodysNightmare**
## Work Package

I see this PR as part of general housekeeping tasks around OpenProject: https://community.openproject.org/work_packages/18714
## Description

Calling `Setting.foo_setting` currently most of the times returns a string in many cases where you actually expect other types to be returned (such as numbers and booleans).

This PR proposes to use the already existing and sporadically used `format` option of the settings to correctly type the return value of `Setting` calls.

I can see this breaking in many places, because currently code is written like this:

```
if Setting.foo_setting == '1'
  call_me Setting.bar_setting.to_i
end
```

But I think I should be worth it long-term-ish. Just want to gather some feedback before adding format information to the whole `settings.yml`.
## Considerations
- stuff will break
- you need to clear the rails cache for changes in return values to become visible

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-typed-settings.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-typed-settings.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by ulferts** [comment]
> Whatever it is, I fear the Greeks even when they bring gifts.

I like the approach but am unsure about the backlash. Would be good for configuration.rb, too.

**Comment by linki** [comment]
:+1: for typed settings (and configuration). the implementation is terrible. (no offense, it fits the rest of the code.) i would seriously consider refactoring that part at some point as settings are a very central aspect. therefore, i strongly suggest you to write some tests for this.

**Comment by NobodysNightmare** [comment]
@linki Any idea on how to do proper tests here?

I don't really want to use existing settings as source for my mock data (e.g. "an int setting should be returned as integer). But since there is no good DI possibility present this seems to be my only choice (apart from doing a larger refactoring of settings in general that would allow for better DI)

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -329,7 +329,7 @@ def denied_auth_source.allow_password_changes?; false; end
     end
 
     it { expect(@u.valid?).to be_falsey }
-    it { expect(@u.errors[:password]).to include I18n.t('activerecord.errors.messages.too_short', count: Setting.password_min_length.to_i) }
+    it { expect(@u.errors[:password]).to include I18n.t('activerecord.errors.messages.too_short', count: Setting.password_min_length) }
```
Comment (NobodysNightmare) [inline spec/models/user_spec.rb:332]: Shut up hound, I already shortened it!

**Comment by ulferts** [comment]
Idle musings: What about the plugins?

File: spec/support/settings.rb [hunk 2]
```diff
@@ -31,9 +31,14 @@
...
 def with_settings(options, &_block)
   saved_settings = options.keys.inject({}) do |h, k|
-    h[k] = Setting[k].is_a?(Symbol) ?
-              Setting[k] :
-              Setting[k].dup
+    begin
+      # try to store a duplicate, but use the original where :dup is not available (int, bool, sym)
+      # apparently this is the only way to determine if dup will work
+      h[k] = Setting[k].dup
+    rescue
```
Comment (linki) [inline spec/support/settings.rb:38]: you can at least define the `dup`-related exception name here

Comment (NobodysNightmare) [inline spec/support/settings.rb:38]: How do I know that is the only one for undupable classes?

**Comment by linki** [comment]
@NobodysNightmare i'm fine with taking some real, stable, obviously typed settings to test the formatting stuff. e.g. app_title for strings, port for integer, some default if `format` is not defined.

just make sure if someone changes the format of one of your setting-under-test she knows quickly what the reason for the failing test is.

File: spec/support/settings.rb (hunk 2 above)
Comment (linki) [inline spec/support/settings.rb:38]: you don't

**Comment by NobodysNightmare** [comment]
I won't have time for this soon-ish, still I think it is a good idea.

Will come back if I get it to a more finished state.

(to the ones able to do so: **don't delete my branch** ;-))

**Pull request closed without merging**